groups/ntc/ntco/ntco_ioring.cpp

```diff
@@ -0,0 +1,6359 @@
+{
+    return d_completionQueue.capacity();
+}
+
+bool IoRingDevice::supportsOperation(
+    ntco::IoRingOperation::Value operation) const
+{
+    return d_probe.isSupported(operation);
+}
+
+bool IoRingDevice::supportsCompletionQueueOverflow() const
+{
+    return d_params.supportsCompletionQueueOverflow();
+}
+
+bool IoRingDevice::supportsEnterTimeout() const
+{
+    return d_params.supportsEnterTimeout();
+}
+
+bool IoRingDevice::supportsNativeWorkers() const
+{
+    return d_params.supportsNativeWorkers();
+}
+
+bool IoRingDevice::supportsCancelByHandle() const
+{
+    return true;
+}
+
+IoRingContext::IoRingContext(ntsa::Handle      handle,
+                             bslma::Allocator* basicAllocator)
+: d_handle(handle)
+, d_pendingEventSetMutex()
+, d_pendingEventSet(basicAllocator)
+, d_allocator_p(bslma::Default::allocator(basicAllocator))
+{
+    BSLS_ASSERT(handle != ntsa::k_INVALID_HANDLE);
+}
+
+IoRingContext::~IoRingContext()
+{
+#if NTCO_IORING_DEBUG
+    if (!d_pendingEventSet.empty()) {
+        for (EventSet::iterator it = d_pendingEventSet.begin();
+             it != d_pendingEventSet.end();
+             ++it)
+        {
+            ntcs::Event* event = *it;
+            NTCO_IORING_LOG_EVENT_PENDING(event);
+        }
+    }
+#endif
+    // We cannot assert that pending events are empty because the this object
+    // is unset as the proactor socket context before this implemenation
+    // asynchronously learns that an event is cancelled. To implement this
+    // properly, ntci::ProactorSocket::cancel must be made asynchronous and
+    // invoke a callback when it is complete.
+
+    // BSLS_ASSERT(d_pendingEventSet.empty());
+}
+
+ntsa::Error IoRingContext::registerEvent(ntcs::Event* event)
+{
+    LockGuard guard(&d_pendingEventSetMutex);
+
+    bool insertResult = d_pendingEventSet.insert(event).second;
+
+    if (!insertResult) {
+        return ntsa::Error::invalid();
+    }
+
+    return ntsa::Error();
+}
+
+void IoRingContext::completeEvent(ntcs::Event* event)
+{
+    LockGuard guard(&d_pendingEventSetMutex);
+
+    bsl::size_t n = d_pendingEventSet.erase(event);
+
+    // Some events may be attempted to be cancelled twice if 'loadPending' is
+    // called quickly in succession and the result set cancelled.
+    //
+    // BSLS_ASSERT(n == 1);
+
+    NTCCFG_WARNING_UNUSED(n);
+}
+
+void IoRingContext::loadPending(EventList* pendingEventList, bool remove)
+{
+    LockGuard guard(&d_pendingEventSetMutex);
+
+    pendingEventList->insert(pendingEventList->end(),
+                             d_pendingEventSet.begin(),
+                             d_pendingEventSet.end());
+
+    if (remove) {
+        d_pendingEventSet.clear();
+    }
+}
+
+ntsa::Handle IoRingContext::handle() const
+{
+    return d_handle;
+}
+
+int IoRingUtil::setup(bsl::size_t entries, ntco::IoRingConfig* parameters)
+{
+    const long k_SYSTEM_CALL_SETUP = 425;
+
+    return static_cast<int>(::syscall(k_SYSTEM_CALL_SETUP,
```

Comment:
I think it can be a good idea to write a comment with the actual system call function name, like `int io_uring_setup(u32 entries, struct io_uring_params *p)` in this case

Reply:
UPD. I got it that there is no glibc wrappers yet for io_uring syscalls, so the comment is not really applicable.

Reply:
OK